drivers/i2s/i2s_ll_stm32.c

```diff
@@ -0,0 +1,1171 @@
+/*
+ * Copyright (c) 2018 STMicroelectronics
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <string.h>
+#include <dma.h>
+#include <i2s.h>
+#include <soc.h>
+#include <clock_control/stm32_clock_control.h>
+#include <clock_control.h>
+
+#include "i2s_ll_stm32.h"
+#include <logging/sys_log.h>
+
+/* FIXME change to
+ * #if __DCACHE_PRESENT == 1
+ * when cache support is added
+ */
+#if 0
+#define DCACHE_INVALIDATE(addr, size) \
+	SCB_InvalidateDCache_by_Addr((u32_t *)addr, size)
+#define DCACHE_CLEAN(addr, size) \
+	SCB_CleanDCache_by_Addr((u32_t *)addr, size)
+#else
+#define DCACHE_INVALIDATE(addr, size) {; }
+#define DCACHE_CLEAN(addr, size) {; }
+#endif
+
+#define MODULO_INC(val, max) { val = (++val < max) ? val : 0; }
+
+static inline unsigned int div_round_closest(u32_t dividend, u32_t divisor)
+{
+	return (dividend + (divisor / 2)) / divisor;
+}
+
+/*
+ * Get data from the queue
+ */
+static int queue_get(struct ring_buf *rb, void **mem_block, size_t *size)
+{
+	unsigned int key;
+
+	key = irq_lock();
+
+	if (rb->tail == rb->head) {
+		/* Ring buffer is empty */
+		irq_unlock(key);
+		return -ENOMEM;
+	}
+
+	*mem_block = rb->buf[rb->tail].mem_block;
+	*size = rb->buf[rb->tail].size;
+	MODULO_INC(rb->tail, rb->len);
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+/*
+ * Put data in the queue
+ */
+static int queue_put(struct ring_buf *rb, void *mem_block, size_t size)
+{
+	u16_t head_next;
+	unsigned int key;
+
+	key = irq_lock();
+
+	head_next = rb->head;
+	MODULO_INC(head_next, rb->len);
+
+	if (head_next == rb->tail) {
+		/* Ring buffer is full */
+		irq_unlock(key);
+		return -ENOMEM;
+	}
+
+	rb->buf[rb->head].mem_block = mem_block;
+	rb->buf[rb->head].size = size;
+	rb->head = head_next;
+
+	irq_unlock(key);
+
+	return 0;
+}
+
+static inline int i2s_stm32_enable_clock(struct device *dev)
```

Comment:
We generally use `static inline` only if we want to place the function in the header file. In this case the `inline` is not required. It's better to let the compiler decide on how to run optimizations. Without `inline` the code is easier to maintain. This function is currently called only once, however it may change in the future.

Using `static inline` in .c file for short functions is probably still OK, it documents the intent for users reading the code.

Reply:
OK, it makes sense.